File: src/main/java/net/database/QuestAPI.java
package net.database;

import client.Character;
import client.player.quest.Quest;
import client.player.quest.QuestState;
import org.jooq.Record;
import org.jooq.Result;

import static database.jooq.Tables.QUESTINFO;
import static database.jooq.Tables.QUESTS;

public class QuestAPI {

    /**
     * Starting a new quest that has never been started before (probably always QuestState.PERFORM)
     *
     * @param quest object
     */
    public static void register(Quest quest) {
        int dbId = DatabaseCore.getConnection()
                .insertInto(QUESTS, QUESTS.QID, QUESTS.CID, QUESTS.STATE)
                .values(quest.getId(), quest.getCharacter().getId(), (byte) quest.getState().getValue())
                .returning(QUESTS.ID)
                .fetchOne().getId();
        quest.setDbId(dbId);

        if (!quest.getMobs().isEmpty()) {
            quest.getMobs().keySet().forEach(mob -> {
                DatabaseCore.getConnection()
                        .insertInto(QUESTINFO, QUESTINFO.QID, QUESTINFO.TYPE, QUESTINFO.KEY, QUESTINFO.VALUE)
                        .values(dbId, (byte) 1, mob, "000")
                        .execute();
            });
        }
    }

    public static void update(Quest quest) {
        DatabaseCore.getConnection()
                .update(QUESTS)
                .set(QUESTS.STATE, (byte) quest.getState().getValue())
                .where(QUESTS.QID.eq(quest.getId()))
                .and(QUESTS.CID.eq(quest.getCharacter().getId()))
                .execute();

        // dont need this data anymore
        if (quest.getState() == QuestState.COMPLETE) {
            DatabaseCore.getConnection()
                    .deleteFrom(QUESTINFO)
                    .where(QUESTINFO.QID.eq(quest.getDbId()))
                    .execute();
        }
    }

    public static void remove(Quest quest) {
        DatabaseCore.getConnection()
                .deleteFrom(QUESTS)
                .where(QUESTS.QID.eq(quest.getId()))
                .and(QUESTS.CID.eq(quest.getCharacter().getId()))
                .execute();
    }

    public static void loadAll(Character chr) {
        Result<Record> res = DatabaseCore.getConnection()
                .select().from(QUESTS)
                .where(QUESTS.CID.eq(chr.getId()))
                .fetch();

        res.forEach(rec -> {
            Quest quest = new Quest(rec.getValue(QUESTS.QID), chr);
            byte state = rec.getValue(QUESTS.STATE);
            if (state == QuestState.PERFORM.getValue()) {
                loadInfo(quest, rec.getValue(QUESTS.ID));
            }
            quest.setState(state == 0 ? QuestState.NONE : (state == 1 ? QuestState.PERFORM : QuestState.COMPLETE));
            quest.setDbId(rec.getValue(QUESTS.ID));
            chr.getQuests().put(quest.getId(), quest);
        });
    }

    private static void loadInfo(Quest quest, int dbId) {
        Result<Record> res = DatabaseCore.getConnection()
                .select().from(QUESTINFO)
                .where(QUESTINFO.QID.eq(dbId))
                .and(QUESTINFO.TYPE.eq((byte) 1))
                .fetch();

        res.forEach(rec -> {
            int mob = rec.getValue(QUESTINFO.KEY);
            quest.getMobs().put(mob, rec.getValue(QUESTINFO.VALUE));
            quest.getCharacter().getRegisteredQuestMobs().add(mob);
        });
    }

    public static void saveInfo(Character chr) {
        chr.getQuests().values().stream()
                .filter(quest -> quest.getState() == QuestState.PERFORM)
                .forEach(quest -> {
                    quest.getMobs().forEach((mob, count) -> {
                        DatabaseCore.getConnection()
                                .update(QUESTINFO)
                                .set(QUESTINFO.VALUE, count)
                                .where(QUESTINFO.KEY.eq(mob))
                                .execute();
                    });
                });
    }
}

File: src/main/java/net/database/QuestAPI.kt
package net.database

import client.Character
import client.player.quest.Quest
import client.player.quest.QuestState
import database.jooq.Tables.QUESTINFO
import database.jooq.Tables.QUESTS
import net.database.DatabaseCore.connection

object QuestAPI {

    /**
     * Starting a new quest that has never been started before (probably always QuestState.PERFORM)
     *
     * @param quest object
     */
    fun register(quest: Quest) {
        with(QUESTS) {
            val dbId = connection.insertInto(this, QID, CID, STATE)
                .values(quest.id, quest.character.id, quest.state.value)
                .returning(ID)
                .fetchOne().id

            quest.dbId = dbId

            with(QUESTINFO) {
                if (quest.mobs.isNotEmpty()) {
                    quest.mobs.keys.forEach {
                        connection.insertInto(this, QID, TYPE, KEY, VALUE)
                            .values(dbId, 1.toByte(), it, "000")
                            .execute()
                    }
                }
            }
        }
    }

    fun update(quest: Quest) {
        with(QUESTS) {
            connection.update(this)
                .set(STATE, quest.state.value)
                .where(QID.eq(quest.id))
                .and(CID.eq(quest.character.id))
                .execute()
        }

        // dont need this data anymore
        with(QUESTINFO) {
            if (quest.state == QuestState.COMPLETE) {
                connection.deleteFrom(this)
                    .where(QID.eq(quest.dbId))
                    .execute()
            }
        }
    }

    fun remove(quest: Quest) {
        with(QUESTS) {
            connection.deleteFrom(this)
                .where(QID.eq(quest.id))
                .and(CID.eq(quest.character.id))
                .execute()
        }
    }

    fun loadAll(chr: Character) {
        with(QUESTS) {
            val res = connection.select().from(this)
                .where(CID.eq(chr.id))
                .fetch()

            res.forEach {
                val quest = Quest(it.getValue(QID), chr)
                val state = it.getValue(STATE)

                if (state == QuestState.PERFORM.value) {
                    loadInfo(quest, it.getValue(ID))
                }

                quest.state =
                    if (state.toInt() == 0) QuestState.NONE else if (state.toInt() == 1) QuestState.PERFORM else QuestState.COMPLETE
                quest.dbId = it.getValue(ID)
                chr.quests[quest.id] = quest
            }
        }
    }

    private fun loadInfo(quest: Quest, dbId: Int) {
        with(QUESTINFO) {
            val res = connection.select().from(this)
                .where(QID.eq(dbId))
                .and(TYPE.eq(1.toByte()))
                .fetch()

            res.forEach {
                val mob = it.getValue(KEY)
                quest.mobs[mob] = it.getValue(VALUE)
                quest.character.registeredQuestMobs.add(mob)
            }
        }
    }

    fun saveInfo(chr: Character) {
        with(QUESTINFO) {
            chr.quests.values
                .filter { it.state == QuestState.PERFORM }
                .forEach { quest ->
                    quest.mobs.forEach {
                        connection.update(this)
                            .set(VALUE, it.value)
                            .where(KEY.eq(it.key))
                            .execute()
                    }
                }
        }
    }
}